File: python/nlhe/range_parser.py
"""Parse poker range strings into combo-weight vectors.

Supported syntax (case-insensitive for ranks, suits lowercase):

    AA              pocket pair, all 6 combos
    AKs             suited hand class, 4 combos
    AKo             offsuit hand class, 12 combos
    AK              suited + offsuit combined, 16 combos
    TT+             pocket pair plus: TT, JJ, QQ, KK, AA
    88-TT           pocket pair range: 88, 99, TT (inclusive)
    A2s+            suited connector-style plus: A2s..AKs
    A5o+            offsuit plus: A5o..AKo
    T9s-76s         descending suited run: T9s, 98s, 87s, 76s
    AhKs            specific combo (explicit cards)
    AA:0.5          weight override (default weight is 1.0)

Multiple tokens are separated by commas or whitespace. Later tokens that mention
the same combo override earlier ones (so you can build a range additively and
carve out exceptions at the end).
"""

from __future__ import annotations

import re
from typing import Dict, List, Tuple

from .cards import (
    INDEX_TO_COMBO,
    NUM_COMBOS,
    RANK_TO_VALUE,
    SUITS,
    VALID_SUITS,
    combo_index,
)

WeightVector = List[float]


class RangeParseError(ValueError):
    pass


_TOKEN_SPLIT = re.compile(r"[,\s]+")


def parse_range(text: str) -> WeightVector:
    weights: Dict[int, float] = {}
    for raw in _TOKEN_SPLIT.split(text.strip()):
        if not raw:
            continue
        token, weight = _split_weight(raw)
        combos = _expand_token(token)
        if not combos:
            raise RangeParseError(f"token matched no combos: {raw!r}")
        for idx in combos:
            weights[idx] = weight

    vec = [0.0] * NUM_COMBOS
    for idx, w in weights.items():
        vec[idx] = w
    return vec


def range_to_combos(weights: WeightVector) -> List[Tuple[Tuple[str, str], float]]:
    from .cards import INDEX_TO_CARD

    out = []
    for idx, w in enumerate(weights):
        if w > 0:
            a, b = INDEX_TO_COMBO[idx]
            out.append(((INDEX_TO_CARD[a], INDEX_TO_CARD[b]), w))
    return out


def range_weight_total(weights: WeightVector) -> float:
    return sum(weights)


def _split_weight(raw: str) -> Tuple[str, float]:
    if ":" in raw:
        token, w = raw.rsplit(":", 1)
        try:
            weight = float(w)
        except ValueError as e:
            raise RangeParseError(f"invalid weight in {raw!r}") from e
        if not 0.0 <= weight <= 1.0:
            raise RangeParseError(f"weight must be in [0, 1], got {weight} in {raw!r}")
        return token.strip(), weight
    return raw.strip(), 1.0


def _expand_token(token: str) -> List[int]:
    if _looks_like_specific_combo(token):
        return [_parse_specific_combo(token)]

    if "-" in token:
        return _expand_dash_range(token)

    if token.endswith("+"):
        return _expand_plus(token[:-1])

    return _expand_class(token)


def _looks_like_specific_combo(token: str) -> bool:
    if len(token) != 4:
        return False
    return token[1] in VALID_SUITS and token[3] in VALID_SUITS


def _parse_specific_combo(token: str) -> int:
    a, b = token[:2], token[2:]
    r1, s1 = a[0].upper(), a[1].lower()
    r2, s2 = b[0].upper(), b[1].lower()
    if r1 not in RANK_TO_VALUE or r2 not in RANK_TO_VALUE:
        raise RangeParseError(f"invalid rank in specific combo {token!r}")
    if s1 not in VALID_SUITS or s2 not in VALID_SUITS:
        raise RangeParseError(f"invalid suit in specific combo {token!r}")
    return combo_index(f"{r1}{s1}", f"{r2}{s2}")


def _expand_class(token: str) -> List[int]:
    """Expand a hand-class token like 'AA', 'AKs', 'AKo', or 'AK'."""
    if len(token) not in (2, 3):
        raise RangeParseError(f"unrecognized token: {token!r}")

    r1 = token[0].upper()
    r2 = token[1].upper()
    if r1 not in RANK_TO_VALUE or r2 not in RANK_TO_VALUE:
        raise RangeParseError(f"invalid rank in token: {token!r}")

    suited = offsuit = pair = False
    if r1 == r2:
        pair = True
        if len(token) != 2:
            raise RangeParseError(f"pair token cannot carry suit marker: {token!r}")
    else:
        if len(token) == 3:
            m = token[2].lower()
            if m == "s":
                suited = True
            elif m == "o":
                offsuit = True
            else:
                raise RangeParseError(f"invalid suit marker: {token!r}")
        else:
            suited = True
            offsuit = True
        if RANK_TO_VALUE[r1] < RANK_TO_VALUE[r2]:
            r1, r2 = r2, r1

    indices: List[int] = []
    if pair:
        for i, s1 in enumerate(SUITS):
            for s2 in SUITS[i + 1 :]:
                indices.append(combo_index(f"{r1}{s1}", f"{r1}{s2}"))
    else:
        for s1 in SUITS:
            for s2 in SUITS:
                if suited and s1 == s2:
                    indices.append(combo_index(f"{r1}{s1}", f"{r2}{s2}"))
                if offsuit and s1 != s2:
                    indices.append(combo_index(f"{r1}{s1}", f"{r2}{s2}"))
    return indices


def _expand_plus(base: str) -> List[int]:
    """Expand 'TT+', 'A2s+', etc. — all hands 'as strong or stronger' in the usual sense."""
    if len(base) == 2 and base[0].upper() == base[1].upper():
        low = RANK_TO_VALUE[base[0].upper()]
        out: List[int] = []
        for v in range(low, 15):
            out.extend(_expand_class(f"{_v2r(v)}{_v2r(v)}"))
        return out

    if len(base) != 3:
        raise RangeParseError(f"invalid '+' token: {base}+")
    high, low, marker = base[0].upper(), base[1].upper(), base[2].lower()
    if high == low:
        raise RangeParseError(f"invalid '+' token (pair cannot have suit marker): {base}+")
    if marker not in ("s", "o"):
        raise RangeParseError(f"invalid suit marker in '+' token: {base}+")
    high_v = RANK_TO_VALUE[high]
    low_v = RANK_TO_VALUE[low]
    if low_v >= high_v:
        raise RangeParseError(f"invalid '+' token (low must be < high): {base}+")

    out = []
    for v in range(low_v, high_v):
        out.extend(_expand_class(f"{high}{_v2r(v)}{marker}"))
    return out


def _expand_dash_range(token: str) -> List[int]:
    left, right = [s.strip() for s in token.split("-", 1)]

    if (
        len(left) == 2
        and len(right) == 2
        and left[0] == left[1]
        and right[0] == right[1]
    ):
        a = RANK_TO_VALUE[left[0].upper()]
        b = RANK_TO_VALUE[right[0].upper()]
        lo, hi = (a, b) if a <= b else (b, a)
        out: List[int] = []
        for v in range(lo, hi + 1):
            out.extend(_expand_class(f"{_v2r(v)}{_v2r(v)}"))
        return out

    if len(left) == 3 and len(right) == 3 and left[2] == right[2] and left[2] in ("s", "o"):
        marker = left[2]
        l_high, l_low = left[0].upper(), left[1].upper()
        r_high, r_low = right[0].upper(), right[1].upper()
        l_gap = RANK_TO_VALUE[l_high] - RANK_TO_VALUE[l_low]
        r_gap = RANK_TO_VALUE[r_high] - RANK_TO_VALUE[r_low]
        if l_gap != r_gap:
            raise RangeParseError(
                f"dash range must preserve gap between ranks: {token!r}"
            )
        high_v_l = RANK_TO_VALUE[l_high]
        high_v_r = RANK_TO_VALUE[r_high]
        lo_high, hi_high = (high_v_r, high_v_l) if high_v_r <= high_v_l else (high_v_l, high_v_r)
        out = []
        for hv in range(lo_high, hi_high + 1):
            lv = hv - l_gap
            out.extend(_expand_class(f"{_v2r(hv)}{_v2r(lv)}{marker}"))
        return out

    raise RangeParseError(f"unrecognized dash range: {token!r}")


def _v2r(v: int) -> str:
    for rank, val in RANK_TO_VALUE.items():
        if val == v:
            return rank
    raise ValueError(f"invalid rank value: {v}")

File: src/nlhe/range_parser.rs
//! Range string parser, matching `python/nlhe/range_parser.py` syntax.
//!
//! Supported tokens (case-insensitive ranks, lowercase suits):
//!   AA              pocket pair (6 combos)
//!   AKs             suited (4 combos)
//!   AKo             offsuit (12 combos)
//!   AK              suited + offsuit (16 combos)
//!   TT+             pair plus: TT,JJ,QQ,KK,AA
//!   88-TT           pair range
//!   A2s+            suited connector plus: A2s..AKs
//!   A5o+            offsuit plus
//!   T9s-76s         descending suited run
//!   AhKs            specific combo (explicit cards)
//!   AA:0.5          weight override (default 1.0)
//!
//! Tokens separated by commas or whitespace. Later tokens for the same combo
//! override earlier ones.

use std::collections::HashMap;

use super::cards::{combo_index, NUM_COMBOS, RANKS, SUITS};

#[derive(Debug, Clone)]
pub struct RangeParseError(pub String);

impl std::fmt::Display for RangeParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "range parse error: {}", self.0)
    }
}
impl std::error::Error for RangeParseError {}

pub fn parse_range(s: &str) -> Result<Vec<f32>, RangeParseError> {
    let mut weights: HashMap<usize, f32> = HashMap::new();

    for raw in s.split(|c: char| c == ',' || c.is_whitespace()) {
        let raw = raw.trim();
        if raw.is_empty() {
            continue;
        }
        let (token, weight) = split_weight(raw)?;
        let combos = expand_token(token)?;
        if combos.is_empty() {
            return Err(RangeParseError(format!("token matched no combos: {raw}")));
        }
        for c in combos {
            weights.insert(c, weight);
        }
    }

    let mut out = vec![0.0f32; NUM_COMBOS];
    for (i, w) in weights {
        out[i] = w;
    }
    Ok(out)
}

fn split_weight(raw: &str) -> Result<(&str, f32), RangeParseError> {
    if let Some(idx) = raw.find(':') {
        let (token, rest) = raw.split_at(idx);
        let weight: f32 = rest[1..]
            .trim()
            .parse()
            .map_err(|_| RangeParseError(format!("bad weight in {raw}")))?;
        Ok((token, weight))
    } else {
        Ok((raw, 1.0))
    }
}

fn rank_value(c: u8) -> Option<u8> {
    let upper = c.to_ascii_uppercase();
    RANKS.iter().position(|&b| b == upper).map(|i| i as u8 + 2)
}

fn suit_idx(c: u8) -> Option<u8> {
    let lower = c.to_ascii_lowercase();
    SUITS.iter().position(|&b| b == lower).map(|i| i as u8)
}

fn card_idx_from(rank_val: u8, suit_idx: u8) -> u8 {
    (rank_val - 2) * 4 + suit_idx
}

fn expand_token(token: &str) -> Result<Vec<usize>, RangeParseError> {
    let bytes = token.as_bytes();
    // Specific combo: 4 chars, two valid cards (e.g. "AhKs")
    if bytes.len() == 4 {
        if let (Some(r1), Some(s1), Some(r2), Some(s2)) = (
            rank_value(bytes[0]),
            suit_idx(bytes[1]),
            rank_value(bytes[2]),
            suit_idx(bytes[3]),
        ) {
            let c1 = card_idx_from(r1, s1);
            let c2 = card_idx_from(r2, s2);
            if c1 == c2 {
                return Err(RangeParseError(format!("duplicate cards in {token}")));
            }
            return Ok(vec![combo_index(c1, c2)]);
        }
    }

    // Range tokens like "88-TT", "T9s-76s"
    if let Some(idx) = token.find('-') {
        let (lo_tok, hi_tok) = token.split_at(idx);
        let hi_tok = &hi_tok[1..];
        return expand_range(lo_tok, hi_tok);
    }

    // Plus tokens: "TT+", "A2s+", "A5o+"
    if let Some(stripped) = token.strip_suffix('+') {
        return expand_plus(stripped);
    }

    // Class tokens: "AA", "AKs", "AKo", "AK"
    expand_class(token)
}

fn expand_class(token: &str) -> Result<Vec<usize>, RangeParseError> {
    let bytes = token.as_bytes();
    if bytes.len() == 2 {
        // Pair "AA" or both ranks unspecified suit "AK" → suited+offsuit
        let r1 = rank_value(bytes[0]);
        let r2 = rank_value(bytes[1]);
        match (r1, r2) {
            (Some(a), Some(b)) if a == b => Ok(pair_combos(a)),
            (Some(a), Some(b)) => {
                let mut out = suited_combos(a, b);
                out.extend(offsuit_combos(a, b));
                Ok(out)
            }
            _ => Err(RangeParseError(format!("bad token: {token}"))),
        }
    } else if bytes.len() == 3 {
        // "AKs" or "AKo"
        let r1 = rank_value(bytes[0]).ok_or_else(|| RangeParseError(format!("bad rank in {token}")))?;
        let r2 = rank_value(bytes[1]).ok_or_else(|| RangeParseError(format!("bad rank in {token}")))?;
        match bytes[2].to_ascii_lowercase() {
            b's' => Ok(suited_combos(r1, r2)),
            b'o' => Ok(offsuit_combos(r1, r2)),
            _ => Err(RangeParseError(format!("expected s or o suffix in {token}"))),
        }
    } else {
        Err(RangeParseError(format!("unrecognized token: {token}")))
    }
}

fn expand_plus(token: &str) -> Result<Vec<usize>, RangeParseError> {
    let bytes = token.as_bytes();
    if bytes.len() == 2 {
        // Pair plus: "TT+" → TT, JJ, ..., AA
        let r1 = rank_value(bytes[0]);
        let r2 = rank_value(bytes[1]);
        if r1 != r2 || r1.is_none() {
            return Err(RangeParseError(format!("expected pair in plus token: {token}")));
        }
        let r = r1.unwrap();
        let mut out = Vec::new();
        for v in r..=14u8 {
            out.extend(pair_combos(v));
        }
        return Ok(out);
    }
    if bytes.len() == 3 {
        // "AKs+" "A2s+" "A5o+"
        let r1 = rank_value(bytes[0]).ok_or_else(|| RangeParseError(format!("bad rank in {token}")))?;
        let r2 = rank_value(bytes[1]).ok_or_else(|| RangeParseError(format!("bad rank in {token}")))?;
        let suffix = bytes[2].to_ascii_lowercase();
        if r1 <= r2 {
            return Err(RangeParseError(format!("expected r1 > r2 in {token}")));
        }
        let mut out = Vec::new();
        for v in r2..r1 {
            match suffix {
                b's' => out.extend(suited_combos(r1, v)),
                b'o' => out.extend(offsuit_combos(r1, v)),
                _ => return Err(RangeParseError(format!("bad suffix in {token}"))),
            }
        }
        return Ok(out);
    }
    Err(RangeParseError(format!("unrecognized plus token: {token}")))
}

fn expand_range(lo_tok: &str, hi_tok: &str) -> Result<Vec<usize>, RangeParseError> {
    let lo = lo_tok.as_bytes();
    let hi = hi_tok.as_bytes();
    // Pair range: "88-TT"
    if lo.len() == 2 && hi.len() == 2 {
        let lo_r = rank_value(lo[0]);
        if rank_value(lo[1]) != lo_r {
            return Err(RangeParseError(format!("expected pair in {lo_tok}")));
        }
        let hi_r = rank_value(hi[0]);
        if rank_value(hi[1]) != hi_r {
            return Err(RangeParseError(format!("expected pair in {hi_tok}")));
        }
        let lo_r = lo_r.unwrap();
        let hi_r = hi_r.unwrap();
        let (lo_r, hi_r) = if lo_r > hi_r { (hi_r, lo_r) } else { (lo_r, hi_r) };
        let mut out = Vec::new();
        for v in lo_r..=hi_r {
            out.extend(pair_combos(v));
        }
        return Ok(out);
    }
    // Suited/offsuit run: "T9s-76s"
    if lo.len() == 3 && hi.len() == 3 && lo[2] == hi[2] {
        let suffix = lo[2].to_ascii_lowercase();
        let lo_a = rank_value(lo[0]).ok_or_else(|| RangeParseError(format!("bad rank")))?;
        let lo_b = rank_value(lo[1]).ok_or_else(|| RangeParseError(format!("bad rank")))?;
        let hi_a = rank_value(hi[0]).ok_or_else(|| RangeParseError(format!("bad rank")))?;
        let hi_b = rank_value(hi[1]).ok_or_else(|| RangeParseError(format!("bad rank")))?;
        // Same gap
        let gap_lo = lo_a as i8 - lo_b as i8;
        let gap_hi = hi_a as i8 - hi_b as i8;
        if gap_lo != gap_hi || gap_lo <= 0 {
            return Err(RangeParseError(format!(
                "gap mismatch or non-positive in {lo_tok}-{hi_tok}"
            )));
        }
        let (start, end) = if lo_b > hi_b { (hi_b, lo_b) } else { (lo_b, hi_b) };
        let mut out = Vec::new();
        for low in start..=end {
            let high = low as i8 + gap_lo;
            if high < 2 || high > 14 {
                continue;
            }
            match suffix {
                b's' => out.extend(suited_combos(high as u8, low)),
                b'o' => out.extend(offsuit_combos(high as u8, low)),
                _ => return Err(RangeParseError(format!("bad suffix in {lo_tok}-{hi_tok}"))),
            }
        }
        return Ok(out);
    }
    Err(RangeParseError(format!("unsupported range token: {lo_tok}-{hi_tok}")))
}

fn pair_combos(rank: u8) -> Vec<usize> {
    let mut out = Vec::new();
    for s1 in 0..4u8 {
        for s2 in (s1 + 1)..4u8 {
            let c1 = card_idx_from(rank, s1);
            let c2 = card_idx_from(rank, s2);
            out.push(combo_index(c1, c2));
        }
    }
    out
}

fn suited_combos(r1: u8, r2: u8) -> Vec<usize> {
    if r1 == r2 {
        return Vec::new();
    }
    let (hi, lo) = if r1 > r2 { (r1, r2) } else { (r2, r1) };
    (0..4u8)
        .map(|s| combo_index(card_idx_from(hi, s), card_idx_from(lo, s)))
        .collect()
}

fn offsuit_combos(r1: u8, r2: u8) -> Vec<usize> {
    if r1 == r2 {
        return Vec::new();
    }
    let (hi, lo) = if r1 > r2 { (r1, r2) } else { (r2, r1) };
    let mut out = Vec::new();
    for s1 in 0..4u8 {
        for s2 in 0..4u8 {
            if s1 == s2 {
                continue;
            }
            out.push(combo_index(card_idx_from(hi, s1), card_idx_from(lo, s2)));
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn count_nonzero(v: &[f32]) -> usize {
        v.iter().filter(|&&x| x > 0.0).count()
    }

    #[test]
    fn pocket_pair_has_six_combos() {
        let v = parse_range("AA").unwrap();
        assert_eq!(count_nonzero(&v), 6);
    }

    #[test]
    fn suited_has_four_combos() {
        let v = parse_range("AKs").unwrap();
        assert_eq!(count_nonzero(&v), 4);
    }

    #[test]
    fn offsuit_has_twelve_combos() {
        let v = parse_range("AKo").unwrap();
        assert_eq!(count_nonzero(&v), 12);
    }

    #[test]
    fn unsuited_token_combines_both() {
        let v = parse_range("AK").unwrap();
        assert_eq!(count_nonzero(&v), 16);
    }

    #[test]
    fn pair_plus() {
        let v = parse_range("TT+").unwrap();
        // TT, JJ, QQ, KK, AA = 5 ranks × 6 combos = 30
        assert_eq!(count_nonzero(&v), 30);
    }

    #[test]
    fn pair_range() {
        let v = parse_range("88-TT").unwrap();
        // 88, 99, TT
        assert_eq!(count_nonzero(&v), 18);
    }

    #[test]
    fn suited_plus() {
        let v = parse_range("A2s+").unwrap();
        // A2s..AKs = 12 hands × 4 combos = 48
        assert_eq!(count_nonzero(&v), 48);
    }

    #[test]
    fn weighted_token() {
        let v = parse_range("AA:0.5").unwrap();
        let nz: Vec<f32> = v.iter().copied().filter(|&x| x > 0.0).collect();
        assert_eq!(nz.len(), 6);
        for w in nz {
            assert!((w - 0.5).abs() < 1e-6);
        }
    }

    #[test]
    fn multi_token() {
        let v = parse_range("AA, KK, AKs").unwrap();
        assert_eq!(count_nonzero(&v), 6 + 6 + 4);
    }

    #[test]
    fn specific_combo() {
        let v = parse_range("AhKs").unwrap();
        assert_eq!(count_nonzero(&v), 1);
    }

    #[test]
    fn descending_suited_run() {
        let v = parse_range("T9s-76s").unwrap();
        // T9s, 98s, 87s, 76s = 4 hands × 4 combos = 16
        assert_eq!(count_nonzero(&v), 16);
    }
}
